### ISSUE TITLE
Support for Passkeys

### ISSUE THREAD

**Issue by alensiljak** [PhilippC/keepass2android#2099]
Is it worth (already) tracking the implementation progress of Passkeys and the possibility for Keepass2Android to act as the storage and the key generator?

Some background information:
- https://fidoalliance.org/passkeys/
- https://developers.google.com/identity/passkeys
- https://android-developers.googleblog.com/2022/10/bringing-passkeys-to-android-and-chrome.html

**Comment by pfak** [comment]
Google has implemented Passkeys on their production websites

https://blog.google/technology/safety-security/the-beginning-of-the-end-of-the-password/

**Comment by zburgermeiszter** [comment]
@PhilippC is this feature planned in a future version?

**Comment by cpetry** [comment]
+1

Would like to stay with KeePass if possible!

**Comment by Kareltje1980** [comment]
This is really getting some traction. There is already some collaboration between other keepass clients on how these should be stored in keepass vaults. 

https://github.com/keepassxreboot/keepassxc/pull/8825

I think it would make sense to take a look at this implementation.

**Comment by boergondier** [comment]
+1

**Comment by juanhs12** [comment]
The latest beta version of KeepassXC have the possibility of to do import and export passkeys.

The pull request with the WebAuthn was accept today, 26 october 2023.

https://snapshot.keepassxc.org/latest/

https://github.com/keepassxreboot/keepassxc/pull/8825

**Comment by darkdragon-001** [comment]
The relevant Android documentations seems to be https://developer.android.com/training/sign-in/credential-provider

Pinging @varjolintu who added the support in KeePassXC.

**Comment by cobexer** [comment]
It looks like this will be backported and released with [KeepassXC 2.7.7](https://github.com/keepassxreboot/keepassxc/milestone/41) just about any time now.

Has there been any development on this so far?

**Comment by Calmquist** [comment]
Bitwarden appears to be adding it as part of the migration from Xamarin to MAUI: https://github.com/bitwarden/mobile/tree/feature/maui-migration-passkeys . It appears to still be in early development, but it looks like Bitwarden is using the credential provider API.

**Comment by jhass** [comment]
Latest Firefox on MacOS supports passkeys now, we could store passkeys via Keepass2Android right now via the QR code method.

**Comment by oxivanisher** [comment]
FYI: As of today, Keepass XC 2.7.7 is officially released and supports passkeys.

**Comment by Kranzes** [comment]
Please do not use the Google passkey library on android as it does not work without Google Play Services and I want to be able to use my YubiKey for this on my degoogled phone.

**Comment by hackerd2501** [comment]
@PhilippC Any plans to implement this ?

**Comment by hackerd2501** [comment]
@PhilippC If I use passkey support on keepassxc, are the keepass databases still compatible between keepassxc and keepass2android, even if you dont implement passkeys ?

**Comment by Calmquist** [comment]
> @PhilippC If I use passkey support on keepassxc, are the keepass databases still compatible between keepassxc and keepass2android, even if you dont implement passkeys ?

I haven't used KeePass2Android, but it shouldn't be a problem. KeePassXC will add some custom fields named KPEX_PASSKEY_*.

**Comment by Ironfist69** [comment]
Bump

**Comment by capi** [comment]
> > @PhilippC If I use passkey support on keepassxc, are the keepass databases still compatible between keepassxc and keepass2android, even if you dont implement passkeys ?
> 
> I haven't used KeePass2Android, but it shouldn't be a problem. KeePassXC will add some custom fields named KPEX_PASSKEY_*.

Can confirm that files from KeepassXC 2.7.7 providing Passkeys can be read and written by Keepass2Android without the passkeys getting in the way. As pointed out, the Passkey's info is stored as special fields in the entries.

It would be great if Keepass2Android would follow the field names used there and hence become compatible with KeepassXC 2.7.7+.

**Comment by PhilippC** [comment]
this clearly is one of the hot candidates for the next feature to be implemented. Unfortunately, I haven't had any time to even read the emails related to the app recently so I can't make any promises about this at the moment.

**Comment by capi** [comment]
Having looked at the keepass2android plugin API, I _think_ that it could be implemented with a plugin, if the missing APIs are still not available from Xamarin. Unfortunately neither do I have much Android development experience nor too much time at the moment, but just as an idea for someone able who'd have time.

**Comment by PhilippC** [comment]
I started to investigate how this could be implemented and tested. 

Unfortunately, I didn't find good ways to test (and use) any implementation (I don't have one yet). Can anybody help here? 

*  it is clear that using Passkeys on Android requires Android 14. I have Android 14.
* In order to test the workflow, you need a password manager app supporting Passkeys. I tried Dashlane, Keyguard and Bitwarden. Only Keyguard seemed to work for me.
* the app requesting the passkey needs to support the Credentials Manager API. I tried Firefox, Chrome and Chrome Canary. While all of those support passkeys, none of them worked with any of the aforementioned password manager apps. (I did enable them in the Android settings for password/passkeys.). Only with Dashlane's Demo App (https://github.com/Dashlane/android-passkey-example) I was able to get the workflow going, but only with Keyguard and not Dashlane or Bitwarden.

I think this whole feature will only be useful if this workflow works with the major Android browsers. Can somebody test this a bit more and maybe help me understand better what's required here?

**Comment by alensiljak** [comment]
As far as I remember, KeePassXC browser extension supports PassKeys.
I'm not sure what is required on Android, unfortunately.

**Comment by ishamf** [comment]
Seems like this is still in development in Firefox: https://bugzilla.mozilla.org/show_bug.cgi?id=1862132

**Comment by 9pr19** [comment]
@PhilippC did you have a look at https://www.passkeys.io/ it seems a good starting point. On Samsung devices the default password manager should be compatible with passkeys https://www.samsung.com/uk/apps/samsung-pass/ . From there I guess you will have to test with multiple browser Chrome or FF (but from my point of view it seems to be still a bit buggy). Hope this can help you.

Thank you for your hard work

**Comment by ishamf** [comment]
I tried out Bitwarden, seems I can only add a GitHub passkey using Samsung internet 😅 Chrome never prompts for Bitwarden when adding the passkey.

After it's added, it can be used in both Chrome and Samsung.

**Comment by Ch4s3r** [comment]
@ishamf For chrome there is a flag in `chrome://flags` to set Passkey support to `Enabled` for Bitwarden to work and you cannot have any passkeys in google password manager AFAIK

**Comment by PhilippC** [comment]
@Ch4s3r  thanks a lot for the hint! That is very helpful! With Bitwarden 2024.4.2 I can then see the workflow in action. That should be a very good starting point!

**Comment by Calmquist** [comment]
I have been able to use passkeys on GrapheneOS and Vanadium with Proton Pass. I originally needed to set the web-authentication-android-credential-management flag to "Enabled for Google Password Manager and 3rd party passkeys", but that flag seems to have been removed and made default since I first tested it.

**Comment by kevinlucasilva** [comment]
@PhilippC, I would like to share a discussion that is taking place within the Bitwarden community about Passkeys:

https://community.bitwarden.com/t/passkeys-support-for-mobile-apps

Take a look at this discussion (if you haven't seen it already), as development on Bitwarden is a little more advanced, and with some issues.

Anyway, I hope they can bring this feature, because the app would be even better than it already is.

THX.

**Comment by SnoopAir** [comment]
Some time has past since the last comment on this feature. Is there any update timeline one when this cool feature will be available? Without it, for me moving forward to using passkeys is a nogo :(

**Comment by frankmill** [comment]
@SnoopAir 

That's exactly how I see it.

**Comment by CueHD** [comment]
> I think this whole feature will only be useful if this workflow works with the major Android browsers.

Android apps are beginning to accept passkeys for authentication. One example is the Fastmail app. I'm not sure if it goes through the Android system webview or not.

**Comment by ieugen** [comment]
I tried KeepassXC passkey with Firefox 128 on Github and it worked.
Would be great to have it working on Android via keepassxc.
Seems like Firefox Android 128 will work with Passkeys 

This explains the status of Passkeys on different version of Android ( ~ minute 6) : https://developer.android.com/courses/pathways/passkeys 
- before Android 14 - only google credential manager supported - but synced with google cloud (did not test)
- Android 14+ - multiple credentials managers at the same time (not sure if I can opt out of google cloud sync for my passkeys)

Also decided to sponsor the project to support this feature :)

**Comment by PetrVladimirov** [comment]
An additional use-case for android apps: Australian ubank is moving its entirely customer base to passkeys with no exceptions.

> Do all ubank customers need to use passkeys to log into their ubank app?
> Yes. Since May 2024, all new customers are already logging into their ubank app using passkeys. From 23 July 2024, existing customers can enrol a passkey via their security settings.
> 
> We aim to have all customers using passkeys as their app log in method by November 2024.

[Source](https://www.ubank.com.au/help/current/app-and-online-banking/passkeys/do-all-ubank-customers-need-to-use-passkeys-to-log-into-their-ubank-app)

**Comment by ICEMANno1** [comment]
> I think this whole feature will only be useful if this workflow works with the major Android browsers. Can somebody test this a bit more and maybe help me understand better what's required here?

I'm not an Android developer or an passkey expert. But from [this](https://developer.android.com/identity/sign-in/credential-provider) documentation I understood 1. all browsers which implement the credentials manager API are supported, keepass2android doesn't need to handle all passkey/webauthn details except for storing the key pair with a relationship to the account and application.

Additionally the new credentials manager API would enable a more seamless integration with browsers supporting it than auto fill (and not only for passkeys).

If I can help in any other way, please always feel free to ask. Thanks for all your great work so far!

**Comment by anttiharju** [comment]
Made a donation to support this feature :)

Strongbox (KeePass for macOS / iOS / iPadOS) already supports passkeys. I hope that keepass2android's implementation will be compatible with the passkeys created by it; their blog post on passkeys may be interesting https://strongbox.reamaze.com/kb/faqs/use-passkeys-with-strongbox:

> **A Note on KeePass Interoperability and other Database Formats**
>
> Not all KeePass apps currently support passkeys. Your passkey data is viewable and can be recovered by opening your database with another KeePass app, but many of these apps will not allow you to create and save new passkeys, or to authenticate with your existing passkeys. Unfortunately, for technical reasons, we cannot support passkeys on KeePass 1 or Password Safe format databases. We recommend migrating to a standard KeePass 2 format database, which should be very straightforward and will provide you with a host of other improvements. Drop us a line if you need some help with that.

![image](https://github.com/user-attachments/assets/f4e2bb17-dcdb-4bfe-8dec-b8d09ecc2bc2)

**Comment by anttiharju** [comment]
Realised I could also provide an example file for Strongbox. The database file contains one entry, with a passkey created at https://webauthn.io. Database password is `potato`

[1] [passkey-from-strongbox.kdbx.zip](https://github.com/user-attachments/files/17355724/passkey-from-strongbox.kdbx.zip)

**Comment by capi** [comment]
@anttiharju It seems that Strongbox follows the way that KeepassXC does it in the desktop version (which would make sense "they" supported it first and other implementations following this would be the best for portability). KeepassXC also does store them as custom fields with the same IDs:
![KeePassXC](https://github.com/user-attachments/assets/7e9fe196-97e3-4199-9af2-1eb78453dbd3)

**Comment by capi** [comment]
@rmueller83 This standards seem related to how to exchange passkeys between credential providers, not so much as for when sharing a kdbx file between different implementations (i.e. because using it on multiple platforms).

**Comment by anttiharju** [comment]
@capi happened to come across [this](https://strongboxsafe.com/updates/passkeys/) blog post from strongbox today. There they mention

> Support for passkeys is coming soon to some major KeePass clients like KeePassXC, with whom we’ve worked to ensure compatibility. **We’re hoping other KeePass clients can take advantage of our trail breaking here.**

**Comment by rbdzyx** [comment]
> I think this whole feature will only be useful if this workflow works with the major Android browsers. Can somebody test this a bit more and maybe help me understand better what's required here?

Not sure if you still want a reply to this, but...

I tried Chrome and Brave on Android 14 (Pixel) and they both support creating and authenticating passkeys. Nothing needs to be enabled.
You can store passkeys in Google password manager or Bitwarden for example. You have to enabled it in _Settings/Passwords and accounts_.
You can test passkeys using
https://webauthn.io/
https://www.passkeys.io/

I have Keepass2Android Offline as my primary selected password manager on Android under _Settings/Passwords and accounts._ Bitwarden is also listed there but disabled, and Google (Google Password Manager, Google Play and Google Wallet) is enabled.
I get prompted to select which Google account available on the phone to store the passkey under. 

It doesn't offer Bitwarden as expected on Android, but it does on WIndows PC where I have the Bitwarden add-on for Chrome.

**Comment by elahn** [comment]
> I feel the need to point out that the best way to get Passkeys implemented isn't to spam this issue, but it's to either help out as requested in [#2099 (comment)](https://github.com/PhilippC/keepass2android/issues/2099#issuecomment-2110501145), or [donate](https://philipp.crocoll.net/donate.php?lang=en&app=keepass2android.keepass2android). The fact that there's a need here has been demonstrated enough, and pinging the author (and everyone subscribed to this issue) isn't gonna help.

I've decided to "spam" the author with donations with the note "passkeys" every time my frustration (at services requiring passkeys or making it the only way to easily login) overflows or I want a progress update.

If enough of us do so, the author could hire someone to add the feature if they're unable to devote the time themselves.

Sorry if this has been answered elsewhere and I missed it or forgot reading it (brain fog); is the author open to pull requests for adding features like this?

**Comment by PhilippC** [comment]
good to know, so I'll better not implement this to keep you making donations 😆 

Joking aside, I'm still working on upgrading the app to .net8 which is required to target api-level 34 which is required to post any updates on Google Play. Until this is finished (and it is a huge amount of work, unfortunately), it doesn't make sense to work on anything else (because I couldn't publish it anyway...)

**Comment by JohnLukeBentley** [comment]
The particular feature of passkey support on keepass2android that it would be desirable to have, and is worth special emphasis, is "Cross Device Authentication (CDA)". 

https://www.passkeycentral.org/introduction-to-passkeys/passkey-types#cross-device-authentication
> Cross-device authentication allows people to sign in on a device (for example, laptop, desktop, or mobile) that does not have a passkey using a second device that does (for example, a mobile device).

See video at https://www.passkeycentral.org/design-guidelines/optional-patterns/cross-device-sign-in

So keepass2android would need to lookup the right passkey (in the locally stored keypass database) given whatever that QR code returns.

It's the lack of this feature in keepass2android (or any other KeypassXC compatible Android App) that prevents me from jumping to passkeys. I don't want to workaround this by, for example, registering additonal passkeys on my work laptop (which may not be possible given central organisational control of the laptop).

**Comment by github3845** [comment]
Apparently, 1password uses the PRF (pseudo-random function) to achieve this https://blog.1password.com/encrypt-data-saved-passkeys/ I think that would be an amazing feature, but I wonder how that would work with multiple devices. So passkey portability to a new device does make sense. How odd that they invented passkeys to work with a secure enclave that requires a Google account and it's bound to a device or a $50 usb stick, and now it seems even Google thinks you should just use a password manager. Which was there all along.


Sadly, I can't code. But I hope it will work in keepass-based Android password managers soon. Thank you for all your work.

**Comment by tsposato** [comment]
> good to know, so I'll better not implement this to keep you making donations 😆
> 
> Joking aside, I'm still working on upgrading the app to .net8 which is required to target api-level 34 which is required to post any updates on Google Play. Until this is finished (and it is a huge amount of work, unfortunately), it doesn't make sense to work on anything else (because I couldn't publish it anyway...)

Thanks for your work mate! Do you know roughly when passkey support will be started on?

**Comment by github3845** [comment]
> Thanks for your work mate! Do you know roughly when passkey support will be started on?

Well, he said after api-level 34 is done, which it seems still hasn't happened, so my personal guess would be that we'll have it in a year.

Your comment and my deduction inspired me to try bitwarden instead, but sadly the keepassxc-created passkeys don't work in it, because it for some reason creates passkey entries a bit differently. (same entry, but seems to require a separate one with correct user name and domain, no idea. Maybe I'm stupid) It also doesn't have TOTP in the free version.

I am getting the feeling that I will have to create one passkey per device and per password manager. 

Basically, I can't wait for keepass2android to get passkey support.

**Comment by wilsonwaters** [comment]
What's the likelihood of this getting implemented? Is there any particular community support you need @PhilippC to make it happen?

Looking through the comments, I think (the epic) #2743 was the blocker and is merged now.

It looks like the Android ecosystem has caught up and Chrome has [full support](https://android-developers.googleblog.com/2025/02/chrome-3p-autofill-services-update.html?m=1#:~:text=In%20October%202024%2C%20we%20announced,autofill%20services%20from%20Chrome%20135.) for 3rd party passkey managers now.

Once done, this would be the only android keepass implementation to handle passkeys which would be really cool and a great alternative to Google's built in/closed ecosystem for passkeys.

**Comment by lefduti** [comment]
So at KeepassDX is also more movement in the direction of Passkeys. https://github.com/Kunzisoft/KeePassDX/issues/1421#issuecomment-3078804868 Perhaps you could use some combined force for both products?

**Comment by github3845** [comment]
It works in KeePassDX!! :D  You can compile it yourself. It took me 3 hours to figure out, but I managed to do it and wrote it down, so that anyone can do it.  https://github.com/Kunzisoft/KeePassDX/issues/1421#issuecomment-3216204251

So now Philipp can just steal their code and implement it tomorrow? :D 

I am so so happy to finally have this.

**Comment by 3Galley** [comment]
@PhilippC Id like to +1 on this request :) 
As a user of Strongbox and Keypass2Android would be very useful to be able to use the passkeys Strongbox created on desktop on mobile 

Thank you for your hard work !

**Comment by segevfiner** [comment]
npm is phasing out TOTP, so without passkey support in Keepass2Android (Including cross-device QR flow hopefully). We will be stuck to switching app or use passkeys with Google Password Manager.

**Comment by outerSpaceToast** [comment]
I support this request. Passkeys are getting important more and more. What is the current status on this topic?

**Comment by sonymaster** [comment]
I am interested in the current plans to provide a passkey function. Is this in development or not?

**Comment by PhilippC** [comment]
Unfortunately I barely find the time to work on this, but I know this is a must-have (and much-wanted) feature.

So I now hired a freelancer to add CredentialProviderService support to the app as a prerequisite. This It's expected to be done by the end of this week - then we can proceed with passkeys.

**Comment by DoctorMcKay** [comment]
This is great news! Just sent a donation your way to help cover the freelancer cost.

**Comment by L235** [comment]
I also just sent in a donation, with many thanks!

**Comment by siccovansas** [comment]
Awesome to hear that work for passkeys in Keepass2Android is moving forward ❤️!

Will this also support WebAuthn PRF (or should that be a separate issue)? WebAuthn PRF is also still an [open isssue](https://github.com/keepassxreboot/keepassxc/issues/11920) for KeePassXC. But some interesting use cases are popping up for PRF, e.g. [a tool](https://github.com/keepassxreboot/keepassxc/issues/11920#issuecomment-3699572337) by Moxie Marlinspike for end-to-end encryption for AI chats.

**Comment by PhilippC** [comment]
Hi all, I know you're waiting for this feature and I'm working on it with priority (as much as my limited time permits). Just wanted to give an update to show that *something* is happening :-) 

<img width="921" height="2048" alt="Image" src="https://github.com/user-attachments/assets/195aa833-0609-4219-9d89-df5a3c34bd77" />

Today, I was able to register a passkey with Keepass2Android, authenticate with it and use that entry across KP2A and KeepassXC 🎉 

There is still a lot of integration work and testing to be done, so please don't expect this to be released very soon. It will take some more weeks, I guess. But it seems to be on a good way!

**Comment by outerSpaceToast** [comment]
> Hi all, I know you're waiting for this feature and I'm working on it with priority (as much as my limited time permits). Just wanted to give an update to show that _something_ is happening :-)
> 
> There is still a lot of integration work and testing to be done, so please don't expect this to be released very soon. It will take some more weeks, I guess. But it seems to be on a good way!

Wow, that sounds great! Keep up the good work!
Thank you very much for this great piece of software 😀

**Comment by J-Jamet** [comment]
I am also eager to see your implementation, as you will not encounter all the problems associated with a purely local implementation. If it helps, there is also the [FIDOK library](https://github.com/BryanJacobs/FIDOk/) by @BryanJacobs, which has done an excellent job, because deserializing webauthn JSON objects should be accessible and better implemented in Google's credential-provider library, but it is not.

**Comment by 3Galley** [comment]
> Hi all, I know you're waiting for this feature and I'm working on it with priority (as much as my limited time permits). Just wanted to give an update to show that _something_ is happening :-)
> <img alt="Image" width="921" height="2000" src="https://private-user-images.githubusercontent.com/5421734/540680116-195aa833-0609-4219-9d89-df5a3c34bd77.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NzE2MjgyOTUsIm5iZiI6MTc3MTYyNzk5NSwicGF0aCI6Ii81NDIxNzM0LzU0MDY4MDExNi0xOTVhYTgzMy0wNjA5LTQyMTktOWQ4OS1kZjVhM2MzNGJkNzcucG5nP1gtQW16LUFsZ29yaXRobT1BV1M0LUhNQUMtU0hBMjU2JlgtQW16LUNyZWRlbnRpYWw9QUtJQVZDT0RZTFNBNTNQUUs0WkElMkYyMDI2MDIyMCUyRnVzLWVhc3QtMSUyRnMzJTJGYXdzNF9yZXF1ZXN0JlgtQW16LURhdGU9MjAyNjAyMjBUMjI1MzE1WiZYLUFtei1FeHBpcmVzPTMwMCZYLUFtei1TaWduYXR1cmU9N2UyYmNmZDE2MjFlZDQ1ZTUyNTVmNzE0MWQzMzk0Y2UwNDkwYjlhY2FkNTgyMzMwNDM0MDhmMjM5NTRiM2ZkNCZYLUFtei1TaWduZWRIZWFkZXJzPWhvc3QifQ.R8OaPSdKVZgqeu5d1m8R3xrKR4FmJfK-QnY-I0zSRX8">
> 
> Today, I was able to register a passkey with Keepass2Android, authenticate with it and use that entry across KP2A and KeepassXC 🎉
> 
> There is still a lot of integration work and testing to be done, so please don't expect this to be released very soon. It will take some more weeks, I guess. But it seems to be on a good way!

Awesome !!!! 
It is awesome to hear that you're making progress on the passkey implementation
I also just sent a donation to chip in with the cost of development and supporting this awesome app :)
Since KeepassDX currently supports passkeys I decided to try it for a couple days, and whereas they have a more polished interface there is a lot missing compared to Keepass2Android. On the other hand, I can confirm that the KeepassDX implementation of passkeys is fully compatible with Strongbox on Mac. I was able to use passkeys both in android and on Mac without any hiccups using KeepassDX. 


Can't wait to see Passkeys fully implemented on Keepass2Android :)

**Comment by tburny** [comment]
Amazing news. Is it possible to add a bounty to this feature?

**Comment by jianboy** [comment]
i see it

**Comment by PhilippC** [comment]
Hi all,
please find https://github.com/PhilippC/keepass2android/releases/tag/v1.16-pre0 as a preview release with experimental passkey support 🎉 . Would be great if you could do some real-world testing, but ⚠️ please don't rely on this to work fully (even though I hope it does)⚠️ - make sure you're not locked out anywhere in case passkey authentication fails!

Just after building the release I found that you might have to disable KP2A as autofill service in the Android settings and reenable it such that it also is detected as credential provider (for passkeys).

For more information, see https://github.com/PhilippC/keepass2android/pull/3228.

I suggest you test this first on https://webauthn.io/ where you can register a mock account.

**Comment by PhilippC** [comment]
@J-Jamet I'm not sure if you will be able to learn much from my code. To be honest, it was rather the other way around :-). Some parts are close to your implementation - because I thoroughly read it and found it to be sound and good work. I wouldn't know how I should implement the interface much differently.

Of course my implementation is still not the same in many aspects (different language, partly different libraries, partly different code organization, different UI integration etc.), but whenever the code was somehow resembling yours, I added the "Derived from KeePassDX" note in the source code headers to comply with GPL v3. In addition to this, would you like to be mentioned in any other way, e.g. in the Changelog or so? Just let me know.

**Comment by VinVel** [comment]
> I suggest you test this first on https://webauthn.io/ where you can register a mock account

I did some testing both using webauthn.io and also a 'real' service such as discord.

For webauthn I was able to register a passkey, but once I saved it it gave me the following error on the website: "An unknown error occurred while talking to the credential manager."

Consequently, when trying to authenticate using the created passkey I got the following error: "That username has no registered credentials".
However the second time I tried it it tried to unlock the database but no passkey was presented, so I cancelled the operation leading to the following error: "The operation either timed out or was not allowed. See: https://www.w3.org/TR/webauthn-2/#sctn-privacy-considerations-client."

On discord when I tried using a passkey that was generated by KeePassXC I was able to select it and authenticate it with biometrics but discord gave the following error: "Entry does not contain passkey data".

As you suggested, I first deactivated and then reactivated KeePass2Android as an autofill service. The phone also recognizes KeePass2Android as a PassKey provider but somewhere during the authentication process something goes wrong.

**Comment by J-Jamet** [comment]
@PhilippC That's pretty cool; I'm glad I could help you with the implementation in KeePass2Android. We learn from each other, and that's always a good thing; it will allow us to expand the use of new standards and show that open-source applications are still in the game. If you'd like to include your inspiration in the changelog, I'd be happy, but it's up to you. The important thing is that we continue to develop community-driven applications. I'll take a look at it, and in any case, congratulations, great work. ₍^. ‿ .^₎⟆

**Comment by brtz2001** [comment]
Hi Philipp,

Thanks a lot for the feature! I have tried it with Linkedin, bahn.de and GitHub with following results.
If I create the passkeys on my desktop (with KeepassXC), keepass2android can use them flawlessly. On the other hand with keepass2android, on all 3 sites I could create the passkeys, but could not use them, neither on keepass2android nor on the desktop, just getting an invalid passkey error at login.

Thanks again :-)

**Comment by akki42** [comment]
First off all, many thanks! 

I have tried out [v1.16-pre0](https://github.com/PhilippC/keepass2android/releases/tag/v1.16-pre0) on [webauthn.io](webauthn.io):

1. Registering new account / passkey with Keepass2Android (Firefox / Chrome): 👍
2. Authenticating with that passkey (Firefox / Chrome on 1st android device): 👍
3. Syncing kdbx including passkey, opening kdbx and recognizing passkey in Keepass2Android on 2nd android device: 👍
4. Syncing kdbx including passkey, opening kdbx and recognizing passkey in KeePassXC (linux desktop): 👍
5. Authenticating with passkey on linux desktop:
     - Firefox: 👎 (KeePassXC pop-up: passkey credentials are shown, button "Authenticate"; after button click: Firefox pop-down: "touch your security key" (or similar, translated), KeePassCX pop-down: "Error: Unknown Passkey-Error") 
     - Chrome: uses CTAP (Client to Authenticator Protocol) 🤔 (KeePassXC pop-up as above;  after "Authenticate" button click: QR code is shown (FIDO:/...); when scanned with barcode scanner on android device, something happens with Google Play Services (it requests Bluetooth access, could be location?, but can be denied), then Keepass2Android shows "use saved passkey for webauthn.io"; after biometric verification for Keepass2Android on tablet, log-in on desktop)

Halfway  successful the other way round:
1. [Registering new account / passkey with KeePassXC (Firefox / Chrome)]
2. [Authenticating with passkey on linux desktop: 👍 (without CTAP)]
3. Syncing kdbx including passkey, opening kdbx in Keepass2Android: 👍 
4. Authenticating with passkey in browser (Firefox / Chrome) on android device: 👎 (Keepass2Android does not show any Passkeys for [webauthn.io](webauthn.io))
5. Changing "URL" from "https://webauthn.io" [as saved by KeePassXC] to "passkey:webauthn.io"
6. Authenticating with passkey in browser (Firefox / Chrome) on android device: 👎
   - Chrome: opening [webauthn.io](webauthn.io), Keepass2Android prompts for unlocking database, then biometric verification, then shows saved passkey as potential credential, but selecting passkey (and another biometric verification) does not immediately login (and does not insert username into field); pressing "Authenticate" button, Passkey can again be selected (or "next"), but only results in "An unknown error occurred while talking to the credential manager."
   - Firefox: nothing automatic happening when opening [webauthn.io](webauthn.io); after manually entering username and pressing "Authenticate", Keepass2Android prompts for unlocking, pressing "unlock" or "next": "choose database", after selecting database, return to unlocking prompt
7. Toggled "Protected Field" (translated: "Geschützes Feld" in Keepass2Android or "schützen" in KeePassXC) for KPEX_PASSKEY_CREDENTIAL_ID to "off" (field not "hidden")
8. Authenticating with passkey in browser on android device:
   - Firefox: 👍 (manually entering username and pressing "Authenticate", Keepass2Android prompts for unlocking, pressing "unlock" or "next": "choose database", after selecting database, Keepass2Android prompts with passkey; successful login after selecting passkey or "next"
   - Chrome: 👎 (as above)

Curious to see that it works for @brtz2001 without any manual changes?!

Further notes:
- no idea why I need to "choose database" in certain situations 
- KeePassXC also "protects" KPEX_PASSKEY_PRIVATE_KEY_PEM and KPEX_PASSKEY_USER_HANDLE; that does not hinder Keepass2Android (so might actually be a good standard setting for Keepass2Android as well?)
- KeePassXC uses numerical values for KPEX_PASSKEY_FLAG_BE and KPEX_PASSKEY_FLAG_BE, while KeePass2Android uses text (true, false)

Happy to provide further information if helpful!

**Comment by brtz2001** [comment]
2 more things:
- Android is actually GrapheneOS with Firefox, and desktop is linux with Firefox
- Weirdly enough, webauthn.io works: registering on Android and opening on desktop, and registering on desktop an opening on Android

**Comment by 3Galley** [comment]
Hi Philipp, 

Awesome progress! 
Thank you for working on this 
I did a bit of testing on Webauthn, Discord and ProtonMail
Im using Mac and Android, on mac im loading the database with Strongbox and on Android 16 your latest v1.16-pre0 on a pixel 7pro

Creating an new entry for webauthn.io worked and allows to log in on android, but not on desktop (mac) 
Creating the entry from Desktop allows to log in on both android and on desktop 
The difference may be on what data is being saved, on the attached screenshots, testuser1256 was created from android, notice from desktop looks like it has no passkey data, but viewed from android it does show the data (weird), although as mentioned before by other testers, the data fields are all unprotected when viewed from android. 
testuser9813 was created from desktop, works on both android and desktop and its passkey data is available and protected from both platforms. 

I also tested logging in to discord app on android and failed with a red font message stating "Credential not discoverable or invalid"after prompting to load the correct entry from the database and accepting the bio-metric. This is an existing entry, was created on desktop long ago and its verified functional. 

Also tested logging in to ProtonMail, and succeeded, after prompting me for bio-metric it logged me into proton with the existing credential created on desktop. I used firefox on mobile to log in to the web portal of proton. 

Any other test you want me to do ? 
attached is the log captured during this operations 

[log.txt](https://github.com/user-attachments/files/27140706/log.txt)
<img width="3072" height="4080" alt="Image" src="https://github.com/user-attachments/assets/bb37f9e4-120e-4276-a1be-67ea511c25bc" />
<img width="3072" height="4080" alt="Image" src="https://github.com/user-attachments/assets/d3ade338-1dfb-4acc-8574-282d41b39d94" />
<img width="1080" height="2340" alt="Image" src="https://github.com/user-attachments/assets/5571e5da-654d-4fdd-aa3d-64ea81ccddc7" />

**Comment by PhilippC** [comment]
thanks a lot for your first feedback! I'll look into it!

**Comment by christantoan** [comment]
Trying to create a Passkey for WhatsApp results in offline error message(??) while creating in Google Password Manager was successful

https://github.com/user-attachments/assets/d1f2c90c-c04a-4e05-8504-58edae1fce1b

Somewhat related to this, I can create a Passkey using KeePassDX but unable to use it for E2EE backup encryption, maybe it will also requires PRF extension when Keepass2Android can use Passkeys in WhatsApp?

<img width="1440" height="3120" alt="Image" src="https://github.com/user-attachments/assets/2d149538-967b-4b77-856e-04cb390d7417" />

**Comment by LewsTherinTelescope** [comment]
🎉 Thanks so much for implementing this (and making the app in general)!

It mostly seemed to work, but with some asterisks which I've detailed below.

- I found while testing with an existing Google sign-in passkey that KP2A didn't recognize it was present when its URL was set to `accounts.google.com` (even though the page _was_ on this subdomain), I needed to change it to just `google.com` and then it worked smoothly. Unsure if this is a general issue with subdomains or if it's some peculiarity of this specific situation.
- I was able to create and use a passkey on both [WebAuthn.io](https://webauthn.io/) and [Discord](https://discord.com), but had a similar issue to the other commenter where when synced to desktop they didn't work there (using KeePassXC, it detected the key but failed after approving; just gave a generic `Unknown passkeys error.` message, so no hints as to where the incompatibility might be unfortunately). I'm not sure which side the bug is on, but either way adding the key on desktop instead did make one that worked on both platforms.
- When I tested a desktop-created passkey for Discord, it seems that if KP2A found any entry with the URL `passkey:discord.com` it wouldn't even try to search for entries with just `discord.com`, so in order to get the desktop-created one to show up I had to manually remove the `passkey:` prefix from the mobile-created one. After I did that, it showed _all_ `discord.com` entries regardless of whether they actually had passkey data or not. I assume this is just how the algorithm is set up to search for things rather than a _bug_ though.
	- The Discord app was also very inconsistent with both login and creation, with it occasionally working but mostly throwing generic errors. However, it worked flawlessly on their website so I think it's just their app being jank again rather than a KP2A issue, but including this for completeness just in case.

**Comment by PhilippC** [comment]
Thanks again for the testing and feedback!

I could reproduce some of the issues and made the following code changes:
* Changed how Private Key data is stored in the entries. This fixed issues e.g. on Discord Web and improved compatibility with KeepassXC
* Made CredentialID, PrivateKey and UserHandle fields protected
* changes BE/BS flags to 0/1 instead of false/true for compatibility with other tools
* removed the "passkey:" prefix in the URL field
* when searching for passkey, KP2A now correctly searches the KPEX_PASSKEY_RELYING_PARTY fields instead of URL

With these changes (and running the latest KeepassXC, which was 2.7.12 in my case), I could register and use passkeys for most of the services you mentioned and also use passkeys across KP2A and KeepassXC. 

However, I wasn't able to use KP2A with Whatsapp nor the Discord App. I want to investigate both further before creating another preview release.

**Comment by PhilippC** [comment]
I was able to fix the issues with Discord and WhatsApp. Please try to confirm that https://github.com/PhilippC/keepass2android/releases/tag/v1.16-pre1 works much better than pre0! Thanks for your testing!

**Comment by christantoan** [comment]
Thank you for the update, I tested it and I can create Passkey on WhatsApp but still unable to test it for login as it automatically verified my device each time.

<img width="1440" height="3120" alt="Image" src="https://github.com/user-attachments/assets/14121c23-7c1c-4904-b0e1-6ca6357b0992" />

But it still cannot be used for encrypting backups.

<img width="1440" height="3120" alt="Image" src="https://github.com/user-attachments/assets/e8879388-c786-40da-a4cf-8e2ef4250218" />

**Comment by VinVel** [comment]
> I was able to fix the issues with Discord and WhatsApp. Please try to confirm that https://github.com/PhilippC/keepass2android/releases/tag/v1.16-pre1 works much better than pre0! Thanks for your testing!

Thank you for the update :). 

I did some testing with Discord. If I try to use the Passkey created by KeePassXC I get the following error: "Anmeldeinformationen nicht auffindbar oder ungültig.".

If I first create a passkey with KeePass2Android and then later try to use it to log in, everything works flawlessly.

I did not yet test, whether you can use the passkey created by KeePass2Android with KeePassXC.

Edit: This seems only to work within the App, not in the Browser (Brave Browser). No Matter which Passkey I pick after authenticating nothing happens, no log-in, no error message.

Edit 2: Generally, when trying to register passkeys in the browser it doesn't work, I can create one, but in the last step where you would usually confirm the passkey, the website seems to have forgotten I was registering a passkey in the first place.

**Comment by haldi4803** [comment]
I guess Firefox on Android does not have Passkey Support? 
Or not with this app?
As it tries to save it via Google as a Device Passkey.

**Comment by voland66** [comment]
@haldi4803 :
You need to go to settings and change the preferred passwords and passkeys service from Google to Keepass2Android for firefox and other apps to start using keepass2android as a passkey provider. (I have not tried keepass2android yet but this is needed for any service, e.g. KeepassDX.)

**Comment by cpms** [comment]
Thanks for making this awesome feature — I’ve been waiting for it for a long time.

I downloaded v1.16-pre1 for testing (also tried v1.16-pre0), but for some reason Passkey authentication is not working at all. Whether I register a new Passkey using KP2A, or register it in advance with KeepassDX and then sync the KDBX file to KP2A, neither method works.

I tested multiple websites (including webauthn.io), and none of them work properly. During registration, I can select KP2A and choose a location to create a new entry, and everything works normally up to that point. However, after adding and returning to the browser, an error occurs (unknown error or something like that).

The same thing happens during authentication — after selecting the credential and returning to the browser, got error.

My device is iQOO12, running Android 16.
Additionally, everything works fine when using KeepassDX.

I recorded some videos:

KP2A for registration:
https://github.com/user-attachments/assets/aecf705d-434f-4f17-b5fd-fa9a51f7a7af
KeepassDX for registration and authentication:
https://github.com/user-attachments/assets/c1397e8f-3b08-41f7-901a-15f86ce2e1d2
KP2A for authentication (using the same DB with KeepassDX):
https://github.com/user-attachments/assets/08bf8ec2-c461-440a-80cf-eab832094547

**Comment by Pihaar** [comment]
I've let claude analyze the PR, comments and compared it with other tools and let it generate a report (since this is an AI, hallucinations are possible; do not take it at face value!): 

### Issue 1: `clientExtensionResults` missing from PublicKeyCredential response

**Severity:** High — Spec violation (`required` per WebAuthn Level 3), likely cause of browser-flow failures

**Symptom:** VinVel (2026-05-11, v1.16-pre1): Passkey registration in browser (Brave) fails — *"in the last step where you would usually confirm the passkey, the website seems to have forgotten I was registering a passkey in the first place"*. KP2A-created passkeys work in native apps but not in the browser. Native app issues (WhatsApp, Discord App) were fixed by the `crossOrigin`/origin fix (commit `3535558`).

This suggests that Android's CredentialManager is more tolerant with the JSON in **native app flows** than in **browser flows**, where the response is validated more strictly.

Commit `e91b69b` deliberately removed `clientExtensionResults` — the code contained `json.Put("clientExtensionResults", new JSONObject()); // TODO credProps`, which was then deleted. Android's CredentialManager does **not** add this field automatically — proven by the Keyguard app (alternative Bitwarden/KeePass client with passkey support), which explicitly includes `clientExtensionResults` as an empty object (`put("clientExtensionResults", buildJsonObject { })`) before returning `CreatePublicKeyCredentialResponse`. Google's official Android sample (`android/identity-samples`, `CredentialProvider/MyVault`) does the same.

**Spec reference ([W3C WebAuthn Level 3, Editor's Draft](https://w3c.github.io/webauthn/)):**

```webidl
dictionary RegistrationResponseJSON {
    required DOMString id;
    required Base64URLString rawId;
    required AuthenticatorAttestationResponseJSON response;
    DOMString authenticatorAttachment;
    required AuthenticationExtensionsClientOutputsJSON clientExtensionResults;
    required DOMString type;
};

dictionary AuthenticationResponseJSON {
    required DOMString id;
    required Base64URLString rawId;
    required AuthenticatorAssertionResponseJSON response;
    DOMString authenticatorAttachment;
    required AuthenticationExtensionsClientOutputsJSON clientExtensionResults;
    required DOMString type;
};
```

Normative text: *"The RegistrationResponseJSON.clientExtensionResults member **MUST** be set to the output of getClientExtensionResults()"*

The type `AuthenticationExtensionsClientOutputsJSON` is an empty dictionary — an empty `{}` is therefore spec-compliant.

**Current code:**

```csharp
// AuthenticatorResponses.cs — FidoPublicKeyCredential.ToJson()
public string ToJson()
{
    var json = new JSONObject();
    json.Put("id", id);
    json.Put("rawId", id);
    json.Put("type", "public-key");
    json.Put("authenticatorAttachment", authenticatorAttachment);
    json.Put("response", response);
    // ← clientExtensionResults MISSING
    return json.ToString();
}
```

KeePassDX, [Keyguard](https://github.com/AChep/keyguard-app), and Google's official Android CredentialProvider sample ([`android/identity-samples`](https://github.com/android/identity-samples/blob/main/CredentialProvider/MyVault/app/src/main/java/com/example/android/authentication/myvault/fido/FidoPublicKeyCredential.kt)) all include this field. The code is used for both registration and authentication — both flows are affected.

**Fix:**
```csharp
json.Put("clientExtensionResults", new JSONObject());
```

---

### Issue 2: KeePassXC-created passkeys don't work in KP2A (Discord App)

**Severity:** Medium — Interoperability issue, possibly identical to Issue 1

**Symptom:** VinVel (2026-05-11): *"If I try to use the Passkey created by KeePassXC I get the following error: 'Anmeldeinformationen nicht auffindbar oder ungültig.'"* — KP2A's own passkeys work in the Discord App but not in the browser. Since browser flows are broken entirely (Issue 1), this may be the same root cause manifesting differently in native apps (error message) vs. browser (silent failure).

**Ruled out:**
- Base64 encoding mismatch — both KeePassXC and KP2A use Base64URL no-padding (`QByteArray::Base64UrlEncoding | OmitTrailingEquals` vs. `Base64Flags.UrlSafe | NoPadding | NoWrap`)
- SEC1 vs. PKCS#8 key format — KeePassXC writes PKCS#8 PEM via `Botan::PKCS8::PEM_encode()`, same as KP2A
- Transport filtering — KP2A does not filter by transports, only by credential ID
- userHandle absent vs. empty — KP2A sends the stored Base64URL string from the KPEX field

**Next steps:** Needs debugging with logs enabled. If fixing Issue 1 (`clientExtensionResults`) resolves this as well, it confirms they share the same root cause.

---

### Issue 3: Strongbox compatibility fields missing

**Severity:** Low — only affects users running Strongbox (iOS/macOS) and KP2A in parallel

**Symptom:** No bug report, but potential incompatibility.

**Cause:** Strongbox has historically used different field names:
- `KPEX_PASSKEY_GENERATED_USER_ID` instead of `KPEX_PASSKEY_CREDENTIAL_ID`
- `KPXC_PASSKEY_USERNAME` instead of `KPEX_PASSKEY_USERNAME`

KeePassXC checks both variants as fallback. KP2A does not.

**Fix:** Add fallback checks in `PasskeyStorage.RetrievePasskey()`:
```csharp
var credentialId = entry.Strings.ReadSafe(FIELD_CREDENTIAL_ID);
if (string.IsNullOrEmpty(credentialId))
    credentialId = entry.Strings.ReadSafe("KPEX_PASSKEY_GENERATED_USER_ID");

var username = entry.Strings.ReadSafe(FIELD_USERNAME);
if (string.IsNullOrEmpty(username))
    username = entry.Strings.ReadSafe("KPXC_PASSKEY_USERNAME");
```

---

### Prioritized fix order

| Prio | Issue | Effort | Impact |
|---|---|---|---|
| **P1** | #1 — Add `clientExtensionResults` | 1 line | Spec compliance; likely cause of browser-flow failures |
| **P2** | #2 — Debug KeePassXC passkeys | Medium | Interoperability with desktop-created passkeys |
| **P3** | #3 — Strongbox fallback | Small | Only relevant for Strongbox users |

---

### Reference: KPEX de-facto standard

There is no formal specification for passkey storage in .kdbx files, but a de-facto standard established via [KeePassXC PR #8825](https://github.com/keepassxreboot/keepassxc/pull/8825) and adopted by 8+ implementations (KeePassXC, KeePassDX, Strongbox, KeePassium, KeeForge, OneKeePass, PassKeeZ, KP2A). The canonical rules:

- **Field names:** `KPEX_PASSKEY_*` prefix (not `KPXC_`)
- **Boolean flags:** `"1"` / `"0"` when writing; also accept `"true"` / `"false"` when reading
- **Credential ID / User Handle:** Base64URL, no padding
- **Private Key:** PKCS#8 PEM (EC P-256 standard, RSA/Ed25519 optional)
- **Relying Party:** Domain string without scheme (e.g. `webauthn.io`, not `https://webauthn.io`)
- **Lookup:** By `KPEX_PASSKEY_RELYING_PARTY` field, not by URL field
- **Tag:** `"Passkey"` on the entry
- **Fallback:** Check `KPEX_PASSKEY_GENERATED_USER_ID` and `KPXC_PASSKEY_USERNAME` as legacy fields

### Reference: Field comparison table (current state after v1.16-pre1, 2026-05-11)

| Field | KeePassXC | KeePassDX | KP2A (PR) | Status |
|---|---|---|---|---|
| `KPEX_PASSKEY_USERNAME` | plain text | plain text | plain text | ✅ |
| `KPEX_PASSKEY_PRIVATE_KEY_PEM` | PEM PKCS#8 | PEM PKCS#8 | PEM PKCS#8 | ✅ |
| `KPEX_PASSKEY_CREDENTIAL_ID` | Base64URL | Base64URL | Base64URL | ✅ |
| `KPEX_PASSKEY_USER_HANDLE` | Base64URL | Base64URL | Base64URL | ✅ |
| `KPEX_PASSKEY_RELYING_PARTY` | Domain | Domain | Domain | ✅ |
| `KPEX_PASSKEY_FLAG_BE` | `"1"` | `"1"`/`"0"` | `"1"`/`"0"` | ✅ |
| `KPEX_PASSKEY_FLAG_BS` | `"1"` | `"1"`/`"0"` | `"1"`/`"0"` | ✅ |
| Entry Tag | `"Passkey"` | `"Passkey"` | `"Passkey"` | ✅ |
| Protected Fields | Yes | Yes | Yes | ✅ |
| RP field search | Yes | Yes | Yes | ✅ |
| `clientDataJSON.crossOrigin` | n/a (not an Android app) | Yes | Yes | ✅ |
| App origin (SigningInfo) | n/a (not an Android app) | Yes | Yes | ✅ |
| `clientExtensionResults` | `{}` (browser extension adds it) | `{}` (empty object) | **missing** | ❌ |
| Strongbox fallback | Yes | — | **missing** | ⚠️ |

**Comment by Stoner60** [comment]
I am a User, who use keepass2android and strongbox on Ipados for the same vaultdb.
I will be happy, if both works for me.
Thanks for implementing passkey on Keepass2Android

**Comment by PhilippC** [comment]
Thanks for your testing and feedback!

@cpms  Can you please report which browser (and which version) you have been testing with? And also, which app version?

@Pihaar , thanks for the report. I added that line for "Issue 1" back, even though I didn't observe any change in behavior. I wasn't aware of the Strongbox fallback in KeepassXC. I added it.

@haldi4803 : Please double check that you use 1.16-pre1. Also, please disable KP2A as credential provider in the Android settings, then re-enable it. It should work with Firefox.

I can reproduce the following two issues, but currently don't have an idea how to solve them as they exist in other apps as well: 

## WhatsApp

@christantoan: I wasn't aware of that feature and thus didn't test. I can confirm it doesn't work. Unfortunately, I didn't find any opensource password manager/credential provider for which it does work (including Google's https://github.com/android/identity-samples/tree/main/CredentialProvider/MyVault). It does work with Google's password manager, but I can't tell what they are doing differently. If anybody finds any open-source tool which works, please let me know.

## Brave-Browser

@VinVel 

I had done my testing mostly in Firefox and there it does work. For Chrome and Brave, I also experience issues. Same as with WhatsApp: I didn't find any other open source credential provider (password manager) that works. If anybody can point me to one, I'm happy to investigate the difference.

**Comment by VinVel** [comment]
> ## Brave-Browser
> [@VinVel](https://github.com/VinVel)
> 
> I had done my testing mostly in Firefox and there it does work. For Chrome and Brave, I also experience issues. Same as with WhatsApp: I didn't find any other open source credential provider (password manager) that works. If anybody can point me to one, I'm happy to investigate the difference.

The only open one that kind of works under the same circumstances is KeePassDX. I tried logging in into Discord with a passkey. If that passkey was created by KeePassXC trying to use it with KeePassDX did not work. However, and that is a strange thing, when trying to use the passkey created by KeePass2Android to log into Discord, KeePassDX could use it without any problems. That means, that the passkeys themselves created by KeePass2Android can be used in the Brave Browser though I am not sure what KeePassDX does different here.

**Comment by VinVel** [comment]
As for other open passkey implementations I did find these ones, though I am unsure how helpful they are: https://git.noisruker.de/Juhu1705/open-passkey-authenticator
https://github.com/awesome-algorand/credential-manager-poc
https://github.com/Dashlane/android-passkey-example
https://github.com/ko-koiwai/MyCredentialManager

**Comment by v3DJG6GL** [comment]
> I tried logging in into Discord with a passkey. If that passkey was created by KeePassXC trying to use it with KeePassDX did not work. However, and that is a strange thing, when trying to use the passkey created by KeePass2Android to log into Discord, KeePassDX could use it without any problems.

Which version of KeePassXC are you using? There was a recent change that fixed a incompatibility issue:
https://github.com/keepassxreboot/keepassxc/pull/13042

Previously, KeePassXC set the `BE` (backup eligibility) and `BS` (backup state) flags to `false`. Almost all password managers set these flags to `true`. This meant that, for some passkeys where the authentication provider was stricter, you could not use them with other KeePass clients if you had initially created them with KeePassXC, and vice versa. I experienced this issue with Authelia, for example.
KeePassDX actually exposed these settings, giving users the ability to change the BE and BS flags for passkeys created with KeePassDX.
Fortunately, the behaviour has been changed in line with other KeePass password managers since the latest KeePassXC release, `v2.7.12`.
https://github.com/keepassxreboot/keepassxc/releases/tag/2.7.12

**Comment by VinVel** [comment]
> Which version of KeePassXC are you using? There was a recent change that fixed a incompatibility issue: [keepassxreboot/keepassxc#13042](https://github.com/keepassxreboot/keepassxc/pull/13042)

I am using v. 2.7.12, though considering I created these passkeys a long time ago, it is possible, that that is the fault.

**Comment by v3DJG6GL** [comment]
> I am using v. 2.7.12, though considering I created these passkeys a long time ago, it is possible, that that is the fault.

I am not sure whether this works or you have to create new Passkeys, but: I see both variables exposed at `Advanced` entry settings: `KPEX_PASSKEY_FLAG_BS` and `KPEX_PASSKEY_FLAG_BE`. You could try to change both of them to `1`
But I suppose  the cleaner approach would be to regenerate these Passkeys...

**Comment by cpms** [comment]
> [@cpms](https://github.com/cpms) Can you please report which browser (and which version) you have been testing with? And also, which app version?

Some additional tests with KP2A 1.16-per1:

Browser:
Edge 148.0.3967.55 (❌) (The browser shown in the video)
Firefox 150.0.3 (✅)
Cromite 145.0.7632.120 (❌)
Chrome 148.0.7778.167 (registration✅ authentication❌)

Native app:
Telegram v12.6.4 (✅)
Nagram X v12.6.4 (✅)

**Comment by PhilippC** [comment]
@VinVel did any of these apps work in a use case where KP2A does not?

**Comment by christantoan** [comment]
@PhilippC, 1Password claims they have open source PRF library but I haven't verified yet as I don't have its subscription. Maybe you can take a look as they offer free subscription for open-source developers.

https://1password.com/blog/encrypt-data-saved-passkeys

**Comment by VinVel** [comment]
> [@VinVel](https://github.com/VinVel) did any of these apps work in a use case where KP2A does not?

I tested the first one. I was able to create and log in with a passkey in Brave Browser. However, it did not work with Discord. I tested it with Canva. Tested the same scenario with Canva with KeePass2Android, did not work. Might just be a Discord issue at that point.

**Comment by canton7** [comment]
For Discord, could it be this issue? https://github.com/bitwarden/clients/issues/14355#issuecomment-3481041444

**Comment by JvandeLocht** [comment]
Some additional tests with KP2A 1.16-per1:

Browser:
Iceraven 2.44.0 (✅)
Vanadium 149.0.7827.48 (❌)

Smartphone: GrapheneOS, Android 16, (Build 2026052401)

Interestingly it is the exact reverse for KeepassDX foss (v4.4.3). There Vanadium works and Iceraven does not. Bitwarden with the vaultwarden backend works for both browsers.

**Comment by pddulfer** [comment]
Thanks for working on this @PhilippC - as a long time KP2A user I am really looking forward to passkey support.

Does the new Android Credential Exchange standard from the June update change things in this space?

https://support.google.com/product-documentation/answer/14343500

**Comment by Ajedi32** [comment]
That's a reference to the Credential Exchange Format (CXF) https://fidoalliance.org/specs/cx/cxf-v1.0-ps-errata-20260309.html#sctn-encoding-considerations

It's just a standard format for storing exported password (and passkey) databases. Maybe something KP2A might want to support in the future, but not strictly necessary for supporting passkeys. (Even if KP2A never supports CXF, as long as *any* tool exists that supports converting to/from the CXF to KBDX it'll be possible to import/export creds between Google Password Manager and KeePass.)

**Comment by haldi4803** [comment]
<img width="865" height="1913" alt="Image" src="https://github.com/user-attachments/assets/c5d5c778-3a66-4037-8ea5-d3c2ad4322ac" />

Successfully created and used a Passkey with the myBMW app. 

Pretty sure the app used Webview for that.

**Comment by AndreKR** [comment]
@PhilippC It would be amazing if you could also support the [KeePassHelper](https://keepass.info/plugins.html#kphelper) format.

Instead of having separate string fields:

| Name | Value |
|--------|--------|
| `KPEX_PASSKEY_CREDENTIAL_ID` | `NEpM0A3NOqZfVDlLuCAYAstR3N3xd5ZwZBELUN0EyuM` |
| `KPEX_PASSKEY_FLAG_BE` | `1` |
| `KPEX_PASSKEY_FLAG_BS` | `0` |
| `KPEX_PASSKEY_PRIVATE_KEY_PEM` | `-----BEGIN PRIVATE KEY----- [...]` |
| `KPEX_PASSKEY_RELYING_PARTY` | `webauthn.io` | 
| ... |  | 

it uses single field with a JSON in it:

| Name | Value |
|--------|--------|
| `KPH: PASSKEY_STORAGE` | `{ "CREDENTIAL_ID": "NEpM0A3NOqZfVDlLuCAYAstR3N3xd5ZwZBELUN0EyuM", "RELYING_PARTY": "webauthn.io", "FLAGS": [ "AT", "UP", "UV", "BE" ], "PRIVATE_KEY_PEM": "-----BEGIN PRIVATE KEY-----\n[...]", ... }`  |

**Comment by kazini** [comment]
> [@PhilippC](https://github.com/PhilippC) It would be amazing if you could also support the [KeePassHelper](https://keepass.info/plugins.html#kphelper) format.
> 
> Instead of having separate string fields:
> 
> Name	Value
> `KPEX_PASSKEY_CREDENTIAL_ID`	`NEpM0A3NOqZfVDlLuCAYAstR3N3xd5ZwZBELUN0EyuM`
> `KPEX_PASSKEY_FLAG_BE`	`1`
> `KPEX_PASSKEY_FLAG_BS`	`0`
> `KPEX_PASSKEY_PRIVATE_KEY_PEM`	`-----BEGIN PRIVATE KEY----- [...]`
> `KPEX_PASSKEY_RELYING_PARTY`	`webauthn.io`
> ...	
> it uses single field with a JSON in it:
> 
> Name	Value
> `KPH: PASSKEY_STORAGE`	`{ "CREDENTIAL_ID": "NEpM0A3NOqZfVDlLuCAYAstR3N3xd5ZwZBELUN0EyuM", "RELYING_PARTY": "webauthn.io", "FLAGS": [ "AT", "UP", "UV", "BE" ], "PRIVATE_KEY_PEM": "-----BEGIN PRIVATE KEY-----\n[...]", ... }`

+1, It would be a lot cleaner and open up future compatibility to use a single string for the key, I believe.

**Comment by navilg** [comment]
There is already a variant of Keepass2Android out in open source with passkey support --> KeepassDX. Switched to that after waiting for this long.

**Comment by AllesMeins** [comment]
> There is already a variant of Keepass2Android out in open source with passkey support --> KeepassDX. Switched to that after waiting for this long.

Of course it's your right to choose the product that suits your needs best - but no need to subtly pressure people that provide software for you for free and in their spare time...

**Comment by navilg** [comment]
> There is already a variant of Keepass2Android out in open source with passkey support --> KeepassDX. Switched to that after waiting for this long.
> Of course it's your right to choose the product that suits your needs best - but no need to subtly pressure people that provide software for you for free and in their spare time...

Can't agree more. I didn't commented to pressure the contributors. It was for anyone who are in need of this feature and can afford to try other app as well. Cheers!

**Comment by sagebind** [comment]
> There is already a variant of Keepass2Android out in open source with passkey support --> KeepassDX. Switched to that after waiting for this long.

Point of clarification: KeePassDX is not a _variant_ of Keepass2Android. It is a different application with a different codebase with a different feature set.

**Comment by dimon222** [comment]
how far is it from merge?

**Comment by schklom** [comment]
FYI @AndreKR @PhilippC the KeePassHelper format has the benefit of being able to store multiple passkeys in a single entry, whereas the KPEX format only supports one Passkey per entry.

I don't have a good usecase in mind, but KeePassHelper supports this so some people might want that.

The choice between both (for saving a new passkey) could be a global setting, or per-entry when creating the passkey.

**Comment by CennoxX** [comment]
The implementation of https://github.com/yusei36/KeePassPasskey/ uses separate string fields. Their solution creates multiple entries, to save multiple passkeys.

**Comment by Steltek** [comment]
> The implementation of https://github.com/yusei36/KeePassPasskey/ uses separate string fields. Their solution creates multiple entries, to save multiple passkeys.

It's using the same KeePassXC format that has already been mentioned.

https://github.com/yusei36/KeePassPasskey/#credential-storage
> Passkeys are stored as standard KeePass entries using [KeePassXC's passkey field format](https://github.com/keepassxreboot/keepassxc):
> [...]

**Comment by shajal0190** [comment]
I am on Version 1.15-r4 but it doesnt show keepass2android passkey options anywhere (like nagram x, redotpay). Am i missing something?
*keepass2android is set as default autofill app, passkey wasnt showing. thats why installed keepassdx to see if passkey is actually working or not. (Keepassdx working)
<img width="1080" height="2217" alt="Image" src="https://github.com/user-attachments/assets/c64e8c35-1400-4c84-887f-51cf66efde1f" />
<img width="1080" height="2216" alt="Image" src="https://github.com/user-attachments/assets/97a8d0dc-1eff-47ea-9c30-cc6ce917e20e" />

**Comment by Steltek** [comment]
> I am on Version 1.15-r4 

Passkey support is only in the 1.16 pre-release versions. ([v1.16](https://github.com/PhilippC/keepass2android/releases/tag/v1.16-pre1))
